### DISCUSSION TITLE
middlewares

### DISCUSSION THREAD

**Discussion by agarwalrohit2503** [serverpod/serverpod#277] [Q&A] [answered]
Is there any concept of middlewares there in serverpod 
or we need to call functions inside the endpoints to do so

**Comment by vlidholt** [comment]
Can you explain a bit more in detail what you want to do?

**Reply by agarwalrohit2503** [reply]
So basically in NodeJS or DartFrog we have concepts like middlewares 
Where we can handle the request verify the authentication token in the header and do anything that we need to do before executing the codes of the actual end point
So do we have this kind of feature in Serverpod as well?

**Reply by ameerfayiz** [reply]
https://github.com/serverpod/serverpod/discussions/1216

**Comment by aragornz325** [comment]
@vlidholt  @agarwalrohit2503 

I need to protect the server routes with an apikey, usually that is done with a middleware. Is there something in serverpod that allows me to do it?

**Reply by vlidholt** [reply, marked as answer]
Yes, Serverpod already has support for authentication. I guess one can say that it uses it's own internal middleware. At some point we may restructure this code to make the different middlewares more configurable (e.g., remove logging or authentication).

**Comment by vanlooverenkoen** [comment]
@vlidholt what is the best way to check if a user has access to a specific resource? 

- when the call first enters our code. (extra database queries if you do the same query in the future as well during the same session?)
- when we are actually accessing the resource and doing some validation there (less database queries. but your code constantly contains check if the resources can be accessed)

In essence I want to validate in the beginning before continueing the rest of the business logic. that way I am always sure the session has access to the resource I need to access.

What are your recommendations here?
